### ISSUE TITLE
How to retrieve document sources when using Agents with memory

### ISSUE THREAD

**Issue by ibagur** [askrella/whatsapp-chatgpt#269]
Dear all,

I am using langchain  0.0.218 and chromadb  0.3.26 over python 3.10.

I am trying to retrieve the document source for a given answer when using Agent Chains with memory over a vectorstore. I have tried many different approaches as suggested here, but so far I have not succeeded:

1.  I first define a QA retrieval chain with sources. This chain is giving me the  output as a dictionary, which contains the `answer` itself plust the name of the main `source` document. If I set the `return_source_documents=True`, I obtain further detail on the paragraphs or chunks of text used, as `source documents`. In any case this seems to work fine by its own:
```
model = 'gpt-3.5-turbo-16k-0613'
temperature = 0
llm = ChatOpenAI(temperature=temperature, model=model)

qa = RetrievalQAWithSourcesChain.from_chain_type(
            llm=llm,
            chain_type="stuff",
            retriever=vectorstore.as_retriever(),
            #return_source_documents=True
        )
```

When running this retrieval chain in its default mode (for simplicity), I get a dictionary with the `answer` and the main `sources`:

```
query = 'What kind of event is the text talking about?'
qa_resp = qa({"question": query}, return_only_outputs=True)
print(qa_resp)

{'answer': 'The text is talking about an earthquake disaster.\n', 'sources': 'Reports/TURKIYE EARTHQUAKE NEEDS AND RESPONSE OVERVIEW_Final.pdf'}
```

2. I then define the tools, which in my case will be only the document retriever QA chain. I define the `func` as a lambda, as it will be returning the output as a dictionary with several elements (the answer and the source document at least)
```
# Define tools
tools = [
    Tool(
        name="My Own Documents QA System",
        #func=qa,
        func=lambda question: qa({"question": question}, return_only_outputs=True),
        description="Useful for when you need to answer questions about the aspects asked. Input may be a partial or fully formed question.",
        #return_direct=True, # in case of using the agent as a router and directly return the result
    )
]
```

3. I create a prompt template, in this case using the wrapper `ZeroShotAgent`, which populates the template with the available tools and adds some format (I could have been done also manually using `PrompTemplate`)
```
# Build prompt template
prefix = """Have a conversation with a human, answering the following questions as best you can based on the context and memory available. 
            You have access to these tools:"""
suffix = """Begin!"

{chat_history}
Question: {input}
{agent_scratchpad}"""

prompt = ZeroShotAgent.create_prompt(
            tools,
            prefix=prefix,
            suffix=suffix,
            input_variables=["input", "chat_history", "agent_scratchpad"],
        )
```

4. I now create a Buffer Memory with a key matching the one defined before in the prompt, and a LLM Chain. Then I create an agent_chain with memory. I am using here AgentExecutor, though I have also tried creating a similar agent chain using the `initialize_agent`with AgentType method, with similar results:

```
# Create Buffer Memory
memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)

# Create LLM Chain
llm_chain = LLMChain(
            llm=llm,
            prompt=prompt,
        )

# Create Agent Executor
agent = ZeroShotAgent(llm_chain=llm_chain, tools=tools, verbose=True)
agent_chain = AgentExecutor.from_agent_and_tools(
                agent=agent, tools=tools, verbose=True, memory=memory
            )
```

I then run this agent_chain in verbose mode:

```
query = "What are the most affected zones?"
result = agent_chain(query)
```

And in the verbose output (I have summarized it), I can see that the agent at some point is retrieving both the `answer` and `sources`, and generating a Final Answer

```
> Entering new  chain...
Thought: I need to find out the most affected zones in the context of the earthquake disaster mentioned in the text.
Action: Turkiye Humanitarian Response QA System
Action Input: "What are the most affected zones in an earthquake disaster?"
Observation: {'answer': 'The most affected zones in the earthquake disaster are.... Additionally, there is significant damage to agricultural production .... \n', 'sources': 'Reports/TURKIYE EARTHQUAKE NEEDS AND RESPONSE OVERVIEW_Final.pdf'}
Thought:I now know the final answer.
Final Answer: The most affected zones in the earthquake disaster mentioned in the text are...

> Finished chain.
```

But when exploring the content of the `result` dictionary variable, the `source` key is not available anywhere. I get the last `input`, the `chat_history`so far, and the `output` from the last query.

```
print(result)

{'input': 'What are the most affected zones?', 'chat_history': [HumanMessage(content='What kind of disaster is the text talking about?', additional_kwargs={}, example=False), AIMessage(content='The text is talking about an earthquake disaster.', additional_kwargs={}, example=False), HumanMessage(content='What are the most affected zones?', additional_kwargs={}, example=False), AIMessage(content='The most affected zones in the earthquake disaster mentioned in the text are...', additional_kwargs={}, example=False)], 'output': 'The most affected zones in the earthquake disaster mentioned in the text are...'}

```

Basically, the agent is working as expected , with a single tool for the moment, but I ignore how could I obtain the `source` key in the final result (or the `source_documents` key in case I decide for a more detailed sources) , so I could eventually retrieve these sources when building a more sophisticated tool.

Any suggestions?

Many thanks.

**Comment by ibagur** [comment]
sorry, this issue was not intended here!
